### PR TITLE
fix statics in externs.

### PR DESCRIPTION
**Pull request by rkirov**

### CODE REVIEW

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 1]
```diff
@@ -256,19 +266,33 @@ public String produceDts(Compiler compiler, Depgraph depgraph) {
...
+      String parentPath = getNamespace(symbol.getName());
       boolean isDefault = isDefaultExport(symbol);
-      boolean isTypedef = symbol.getType() != null && isTypedef(symbol.getType());
-      // Typedef symbols do not enumerate their namespace properties.
-      if (!isDefault && !isTypedef) visitedNamespaces.add(symbol.getName());
-      if (isDefault && visitedNamespaces.contains(namespace)) continue;
-      declareNamespace(isDefault ? namespace : symbol.getName(), symbol, isDefault, compiler, transitiveProvides, true);
+
+      if (visitedNamespaces.contains(parentPath)) {
+        // before we continue we need to account for the fact that the class has been visited.
```
Comment (mprobst) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: Maybe formulate the comment as `// Remember that this class has been visited before continuing.`?

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 2]
```diff
@@ -256,19 +266,33 @@ public String produceDts(Compiler compiler, Depgraph depgraph) {
...
+        if (isDefault && isClassLike(symbol.getType())) visitedClassLikes.add(symbol.getName());
+        continue;
+      }
+      if (isStatic(symbol.getType()) && visitedClassLikes.contains(parentPath)) continue;
```
Comment (mprobst) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: add a comment a la `will be handled by XXX` for this case?

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 3]
```diff
@@ -256,19 +266,33 @@ public String produceDts(Compiler compiler, Depgraph depgraph) {
...
+
       // we do not declare modules or goog.require support, because externs types should not be
       // visible from TS code.
     }
+  }
 
-    checkState(indent == 0, "indent must be zero after printing, but is %s", indent);
-    return out.toString();
+  private boolean isClassLike(JSType type) {
+    return isTrueConstructor(type) || type.isInterface();
```
Comment (mprobst) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: do we need the difference between these two (for empty ns?). If so, please comment.

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 4]
```diff
@@ -256,19 +266,33 @@ public String produceDts(Compiler compiler, Depgraph depgraph) {
...
+  }
+
+  private boolean isStatic(JSType type) {
```
Comment (mprobst) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: This needs a better name, it doesn't match the common definition of static. Maybe `isStaticField` or so? I.e. as opposed to a nested ctor?

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 5]
```diff
@@ -392,7 +416,11 @@ private boolean isLikelyNamespace(JSType type, Compiler compiler) {
     if (utype != null) {
       return isNativeObjectType(utype.restrictByNotNullOrUndefined(), compiler);
     }
-    return isNativeObjectType(type, compiler);
+    // The inferred type for an namespace-like object is not the native object,
+    // but rather PrototypeObjectType.
+    return isNativeObjectType(type, compiler) ||
+        // TODO(rado): replace with something shorter that makes all tests pass.
```
Comment (mprobst) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: is this done?

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java (hunk 1 above)
Comment (rkirov) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: done

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java (hunk 2 above)
Comment (rkirov) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: done.

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java (hunk 3 above)
Comment (rkirov) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: The name was confusing, renamed to 'allowsStaticFields' and indeed we do need to check for them separately, interfaces are not constructors.

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java (hunk 4 above)
Comment (rkirov) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: renamed to isStaticFieldorMethod.

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java (hunk 5 above)
Comment (rkirov) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: I give up, can't find a better API to check for PrototypeObjectType. But I want to leave a comment to show my aspirations for a better api :) Changed the comment a bit.

**Comment by rkirov** [comment]
PTAL

**Comment by mprobst** [comment]
Needs better commit message, LGTM otherwise.

**Pull request merged**